From 7adcd01d96ac8f9077d32f214c85c0d8ef866308 Mon Sep 17 00:00:00 2001
From: Mats Erik Andersson <gnu@gisladisker.se>
Date: Thu, 8 Oct 2015 19:04:43 +0200
Subject: [PATCH 15/42] syslogd: Redefine faulty macro LOG_MAKEPRI.

Old versions of Glibc cause remotely logged kernel
messages to be rewritten for facility `uucp', not
the intended facility `user'.
---
 ChangeLog     | 15 +++++++++++++++
 src/logprio.h | 12 ++++++++++++
 src/syslogd.c | 13 +++++++++++++
 3 files changed, 40 insertions(+)

diff --git a/src/logprio.h b/src/logprio.h
index 79072e02..315ad052 100644
--- a/src/logprio.h
+++ b/src/logprio.h
@@ -66,6 +66,18 @@ typedef struct _code {
 #  define LOG_PRI(p)	((p) & LOG_PRIMASK)
 #endif
 
+/* Glibc prior to 2.17 included a definition of LOG_MAKEPRI
+ * that evaluated LOG_MAKEPRI(LOG_USER, 0) to (1 << 9).
+ * The first argument was shifted three bits, ignoring
+ * the definition LOG_USER = (1 << 3).  Avoid this
+ * harmful mistake.
+ */
+#ifdef LOG_MAKEPRI
+# if LOG_MAKEPRI (1, 0) > LOG_PRIMASK
+#  undef LOG_MAKEPRI
+# endif
+#endif /* LOG_MAKEPRI */
+
 #ifndef LOG_MAKEPRI
 #  define LOG_MAKEPRI(fac, p)	((fac) | (p))
 #endif
diff --git a/src/syslogd.c b/src/syslogd.c
index 0fa41945..352a5bd0 100644
--- a/src/syslogd.c
+++ b/src/syslogd.c
@@ -116,6 +116,19 @@
 # include "logprio.h"
 #endif
 
+/* Glibc prior to 2.17 included a definition of LOG_MAKEPRI
+ * that evaluated LOG_MAKEPRI(LOG_USER, 0) to (1 << 9).
+ * The first argument was shifted three bits, ignoring
+ * the definition LOG_USER = (1 << 3).  Avoid this
+ * harmful mistake.
+ */
+#ifdef LOG_MAKEPRI
+# if LOG_MAKEPRI (1, 0) > LOG_PRIMASK
+#  warning Discarding faulty LOG_MAKEPRI defined in system header file.
+#  undef LOG_MAKEPRI
+# endif
+#endif /* LOG_MAKEPRI */
+
 #ifndef LOG_MAKEPRI
 #  define LOG_MAKEPRI(fac, p)	((fac) | (p))
 #endif
-- 
2.20.1.791.gb4d0f1c61a

